proposals/NNNN-reflection.md

```diff
@@ -0,0 +1,704 @@
+
+We can take each of the motivating examples and see how we can achieve our
+result using these new `Reflection` APIs.
+
+Our registration example can finally iterate a type's fields without even
+having access to an instance of the type by using the new `Type` API:
+
+```swift
+import Reflection
+
+func register(_ type: Any.Type) {
+  for prop in Type(type).storedProperties {
+    if prop.type.swiftType is CustomRegister.Type {
+      cache[type].append(prop)
+    }
+  }
+}
+```
+
+In our enum case name example, we can use a new type called `Case` and use our
+enum instance to initialize a value of this `Case` and grab the name out:
+
+```swift
+import Reflection
+
+enum Food: String, CodingKey {
+  case burger = "Hamburger"
+  case fries = "French Fries"
+  case tots = "Tater tots"
+  case milkshake = "Milkshake"
+}
+
+let favoriteFood = Food.tots
+let caze = Case(from: favoriteFood)!
+
+print(caze.name) // "tots"
+```
+
+Finally, we can use a new type called `PartialType` that lets us do these sort
+of generic erased queries and utilize another new API on `Type` called
+`genericArguments`:
+
+```swift
+import Reflection
+
+func sequenceAlgorithm<T: Sequence>(_ seq: T) {
+  let genericArgs = Type(T.self).genericArguments
+
+  // Void doesn't matter here as you'll see later on
+  if Type(T.self).partial == Type([Void].self).partial {
+    let elementType = genericArgs[0]
+  }
+
+  // This dictionary type doesn't matter here as you'll see later on
+  if Type(T.self).partial == Type([AnyHashable: Any].Keys.self).partial {
+    let keyType = genericArgs[0]
+    let valueType = genericArgs[1]
+  }
+
+  // Do slow thing
+}
+```
+
+## Detailed design
+
+`Reflection` will be a new module included in the Swift toolchain. In order to
+use this module, one must explicitly declare an `import Reflection`.
+
+### `Type`
+
+This is our main entry point into the reflection module. `Type` represents all
+possible concrete types that can occur in Swift. Ranging from structs, enums,
+tuples, functions, existentials, etc. Anything that can be on the right hand
+side of an `as` is represented here in `Type`. 
+
+```swift
+@frozen
+public struct Type {
+  @usableFromInline
+  let metadata: Metadata
+
+  /// Converts a Swift metatype to a 'Type'.
+  @inlinable
+  public init(_ type: Any.Type)
+
+  /// Gets the runtime type of the instance and makes a 'Type'.
+  @inlinable
+  public init(_ instance: Any)
+}
+
+extension Type {
+  /// Returns a collection wrapper over 'Case'
+  @inlinable
+  public var cases: Cases { get }
+
+  /// Returns a collection wrapper over 'StoredProperty'
+  @inlinable
+  public var storedProperties: StoredProperties { get }
+
+  /// Returns a collection wrapper over the generic arguments needed to
+  /// construct this concrete 'Type'. The collection will always be empty for
+  /// types who are not structs, classes, or enums as well as types who are not
+  /// generic.
+  @inlinable
+  public var genericArguments: GenericArguments { get }
```

Comment:
I don't see a description of the `GenericArguments` type in the proposal. Could you please include more details about this type, such as what its `Element` type is? I assume the elements are also `Type`s but that may not be obvious to a reviewer.